04.ls/ls.rb

```diff
@@ -2,33 +2,92 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'etc'
 
 COL_MAX = 3
 PADDING = 2
+FILE_TYPE_TO_CHARACTER = {
+  'file' => '-',
+  'directory' => 'd',
+  'characterSpecial' => 'c',
+  'blockSpecial' => 'b',
+  'fifo' => 'p',
+  'link' => 'l',
+  'socket' => 's',
+  'unknown' => '?'
+}.freeze
+PERMISSION_SEPARATER = 512
 
 def main
-  options = { r: false }
+  options = { l: false }
   opt = OptionParser.new
-  opt.on('-r') { options[:r] = true }
+  opt.on('-l') { options[:l] = true }
   argv = opt.parse(ARGV)
 
-  path = argv[0] || '.'
-  FileTest.directory?(path) or return
-  file_names = Dir.children(path).sort
-  display_file_names(file_names, options)
+  path = argv[0] || './'
+  corrected_path = path.end_with?('/') ? path : "#{path}/"
+  FileTest.directory?(corrected_path) or return
+  file_names = Dir.children(corrected_path).sort
+  filtered_file_names = file_names.reject { |name| name.start_with?('.') }
+  options[:l] ? display_file_names_with_long(corrected_path, filtered_file_names) : display_file_names(filtered_file_names)
 end
 
-def display_file_names(file_names, options)
-  filtered_file_names = file_names.reject { |name| name.start_with?('.') }
-  file_names_to_display = options[:r] ? filtered_file_names.reverse : filtered_file_names
-  total_file_names = file_names_to_display.size.to_f
+def display_file_names_with_long(path, file_names)
+  file_paths = file_names.map { |file_name| path + file_name }
+  total_block_size = file_paths.sum { |file_path| File.stat(file_path).blocks }
+  column_name_to_width = get_column_name_to_width(file_paths)
+  puts "total #{total_block_size / 2}"
+
+  file_paths.each do |file_path|
+    file_type = File.ftype(file_path)
+    stat = File.stat(file_path)
+    mode_number = stat.mode % PERMISSION_SEPARATER
+    mode = convert_mode_number_to_mode(mode_number)
```

Comment:
```suggestion
    mode = convert_mode_number_to_mode(stat)
```

みたいにした方がスッキリして良さそうです。（その場合メソッド名は再検討が必要ですが）